padertorch/train/hooks.py

```diff
@@ -812,105 +816,115 @@ class StopTraining(Exception):
     pass
 
 
-class LossWeightAnnealingHook(TriggeredHook):
-    """
-    Anneals a loss weight within the los_weights dict of the trainer.
-    """
-    def __init__(self, name, factor, trigger, max_value=None, min_value=None):
+class AnnealingHook(TriggeredHook):
+    def __init__(self, trigger, breakpoints, name):
         """
+        Base class for piece-wise linear annealing. The piece-wise linear
+        function is parameterized by its breakpoints. Can also be used for
+        arbitrary annealing functions by stating breakpoints with an interval
+        similar to the trigger interval.
 
         Args:
-            name: key of the loss_weight
-            factor: factor by which to anneal the loss weight.
-                factor > 1. results in an increase while factor < 1. results
-                in a decrease
             trigger:
-            max_value: upper bound of the weight
-            min_value: lower bound of the weight
-                (hint: can also be used to activate a loss weight after a
-                certain number of iterations/epochs)
+            breakpoints: list of (iteration, value) coordinates of the piecewise linear funtion.
+            name: name of the attribute. You can use "attr1.attr11" to
+                anneal a sub attribute
         """
         super().__init__(trigger)
+        self.breakpoints = sorted(breakpoints, key=lambda x: x[0])
         self.name = name
-        self.factor = factor
-        self.max_value = max_value
-        self.min_value = min_value
+        self.last_break = None
 
     @property
     def uid(self):
-        return super().uid + f"({self.name})"
+        return super().uid + f"({'.'.join(self.name)})"
+
+    def get_value(self, trainer):
+        raise NotImplementedError
+
+    def set_value(self, trainer, value):
+        raise NotImplementedError
 
     def pre_step(self, trainer):
-        if self.trigger(iteration=trainer.iteration, epoch=trainer.epoch) \
-                and trainer.iteration != 0:
-            weight = self.factor * trainer.loss_weights[self.name]
-            if self.max_value is not None:
-                weight = min(weight, self.max_value)
-            if self.min_value is not None:
-                weight = max(weight, self.min_value)
-            trainer.loss_weights[self.name] = weight
+        if self.trigger(iteration=trainer.iteration, epoch=trainer.epoch):
+            if self.last_break is None:
+                self.last_break = (0, self.get_value(trainer))
+            while len(self.breakpoints) > 0 and self.breakpoints[0][0] <= trainer.iteration:
+                self.last_break = self.breakpoints.pop(0)
+            if len(self.breakpoints) > 0:
+                slope = (self.breakpoints[0][1]-self.last_break[1])/(self.breakpoints[0][0]-self.last_break[0])
+                value = self.last_break[1] + slope * (trainer.iteration - self.last_break[0])
+            else:
+                value = self.last_break[1]
```

Comment:
This code is relative complicated.
Is my understanding correct: Take the original value as start values and then interpolate to the next break point?
i.e. the breakpoints are absolute and not relative the original value.

Reply:
right, I added comments to improve clarity of the math